### DISCUSSION TITLE
websockets.legacy is deprecated and now causes the tests to fail

### DISCUSSION THREAD

**Discussion by juliangilbey** [Kludex/uvicorn#2530] [Potential Issue] [answered]
As reported in Debian at https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1088725 the current release of websockets (14.1) reports a `DeprecationWarning` for `websockets.legacy`, which causes the package tests to fail.  This could be ignored by adding a suitable line to `pyproject.toml`, or the usage could be replaced.  The test run gives:

```
============================= test session starts ==============================
platform linux -- Python 3.13.0, pytest-8.3.3, pluggy-1.5.0 -- /usr/bin/python3.13
cachedir: .pytest_cache
rootdir: /<<PKGBUILDDIR>>/.pybuild/cpython3_3.13/build
configfile: pyproject.toml
plugins: asyncio-0.24.0a1, mock-3.14.0, typeguard-4.4.1, anyio-4.6.2
asyncio: mode=Mode.STRICT, default_loop_scope=None
collecting ... collected 621 items / 1 error / 40 deselected / 581 selected

==================================== ERRORS ====================================
________________ ERROR collecting tests/test_auto_detection.py _________________
tests/test_auto_detection.py:10: in <module>
    from uvicorn.protocols.websockets.auto import AutoWebSocketsProtocol
uvicorn/protocols/websockets/auto.py:19: in <module>
    from uvicorn.protocols.websockets.websockets_impl import WebSocketProtocol
uvicorn/protocols/websockets/websockets_impl.py:10: in <module>
    import websockets.legacy.handshake
/usr/lib/python3/dist-packages/websockets/legacy/__init__.py:6: in <module>
    warnings.warn(  # deprecated in 14.0 - 2024-11-09
E   DeprecationWarning: websockets.legacy is deprecated; see https://websockets.readthedocs.io/en/stable/howto/upgrade.html for upgrade instructions
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
======================= 40 deselected, 1 error in 0.44s ========================
```

**Comment by Kludex** [comment, marked as answer]
PR welcome to add the line to pyproject to help you folks.

**Reply by juliangilbey** [reply]
Given #1908, I won't submit a PR for this; that route is far better.

**Comment by juliangilbey** [comment]
Of course that will help for the time being (till 2030!), but when the functionality is removed entirely, this code will stop working. The upgrade path is described here: https://websockets.readthedocs.io/en/stable/howto/upgrade.html but I don't know enough about uvicorn to propose a PR to migrate to the new asyncio version of websockets.

**Reply by Kludex** [reply]
Please search for similar issues/discussions before creating a new discussion: https://github.com/encode/uvicorn/issues/1908

**Reply by juliangilbey** [reply]
Oh, sorry, I missed that in my search.

**Comment by juliangilbey** [comment]
Turns out that that issue does not cover all of the deprecation warnings; I'm not sure whether I should add a note of those here, or to that issue, or open a separate discussion?

They are:

* `websockets.server.WebSocketServerProtocol` in `uvicorn/protocols/websockets/websockets_impl.py`
* `websockets.client.connect` in `tests/middleware/test_proxy_headers.py`, `tests/middleware/test_logging.py`, `tests/protocols/test_websocket.py` (lots of times)
